model/extensions/V/vext_vset_insts.sail

```diff
@@ -77,47 +76,50 @@ function calculate_new_vl(AVL : xlenbits, VLMAX : range(1, vlen)) -> range(0, vl
   } else VLMAX
 }
 
-// ********************************** vsetvli ***********************************
-union clause instruction = VSETVLI : (bits(1), bits(1), bits(3), bits(3), regidx, regidx)
-
-mapping clause encdec = VSETVLI(ma, ta, sew, lmul, rs1, rd)
-  <-> 0b0000 @ ma @ ta @ sew @ lmul @ encdec_reg(rs1) @ 0b111 @ encdec_reg(rd) @ 0b1010111
-  when currentlyEnabled(Ext_V)
-
-function clause execute VSETVLI(ma, ta, sew, lmul, rs1, rd) = {
-  let LMUL_pow_ori  = get_lmul_pow();
-  let SEW_pow_ori   = get_sew_pow();
-  let ratio_pow_ori = SEW_pow_ori - LMUL_pow_ori;
-
-  // check new SEW and LMUL are legal and calculate VLMAX
+// Execute vsetvl, vsetvli, or vsetivli.
+function execute_vsetvl_type(
+  ma : bits(1),
+  ta : bits(1),
+  sew : bits(3),
+  lmul : bits(3),
+  avl : xlenbits,
+  requires_fixed_vlmax : bool,
+  rd : regidx,
+) -> ExecutionResult = {
+  // Check new SEW and LMUL have legal encodings. There are some reserved values.
   if is_invalid_lmul_pow(lmul) | is_invalid_sew_pow(sew) then {
     handle_illegal_vtype();
-    return RETIRE_SUCCESS
+    return RETIRE_SUCCESS;
   };
+
   let LMUL_pow_new = lmul_pow_val(lmul);
   let SEW_pow_new = sew_pow_val(sew);
 
-  if SEW_pow_new > (LMUL_pow_new + elen_exp) then { handle_illegal_vtype(); return RETIRE_SUCCESS };
-  let VLMAX = 2 ^ (LMUL_pow_new + vlen_exp - SEW_pow_new);
+  let lmul_sew_ratio = get_lmul_pow() - get_sew_pow();
+  let lmul_sew_ratio_new = LMUL_pow_new - SEW_pow_new;
 
-  // Set vl according to VLMAX and AVL.
-  if rs1 != zreg then {
-    // Normal stripmining.
-    vl = to_bits(calculate_new_vl(X(rs1), VLMAX));
-    X(rd) = vl;
-  } else if rd != zreg then {
-    // Set vl to VLMAX.
-    vl = to_bits(xlen, VLMAX);
-    X(rd) = vl;
-  } else {
-    // Keep existing vl if it is compatible and vtype.ill is valid.
-    let ratio_pow_new = SEW_pow_new - LMUL_pow_new;
-    let is_legal_vset = ratio_pow_new == ratio_pow_ori & valid_vtype();
-    if not(is_legal_vset) then { handle_illegal_vtype(); return RETIRE_SUCCESS }
+  // "For a given supported fractional LMUL setting, implementations must
+  // support SEW settings between SEWMIN and LMUL * ELEN, inclusive."
+  //
+  // Also some cases require that VLMAX doesn't change (i.e. LMUL/SEW is constant).
+  if SEW_pow_new > (LMUL_pow_new + elen_exp) |
+     (requires_fixed_vlmax & (lmul_sew_ratio != lmul_sew_ratio_new)) then {
```

Comment:
The `requires_fixed_vlmax` case should also ensure: "Use of the instructions
is also reserved if `vill` was 1 beforehand."
```suggestion
     (requires_fixed_vlmax & (lmul_sew_ratio != lmul_sew_ratio_new | not(valid_vtype())) then {
```

Reply:
Hmm this is handled on line 165, but it looks like it was actually missing for `vsetvli` before, even though the comment says `and vtype.ill is valid` (unless there's a check I'm not seeing).

It probably makes sense to handle it here anyway.